master/internal/core.go

```diff
@@ -146,6 +149,65 @@ func (m *Master) getMasterLogs(c echo.Context) (interface{}, error) {
 	return entries, nil
 }
 
+func (m *Master) getResourceAllocationRaw(c echo.Context) error {
+	args := struct {
+		AfterDate  string `query:"after_date"`
+		BeforeDate string `query:"before_date"`
+	}{}
+	if err := api.BindArgs(&args, c); err != nil {
+		return err
+	}
+
+	afterDate, err := time.Parse("2006-01-02", args.AfterDate)
+	if err != nil {
+		return errors.Wrap(err, "invalid after date")
+	}
+	beforeDate, err := time.Parse("2006-01-02", args.BeforeDate)
+	if err != nil {
+		return errors.Wrap(err, "invalid before date")
+	}
+	if afterDate.After(beforeDate) {
+		return errors.New("the after date cannot be after the before date")
+	}
+
+	resp := &apiv1.ResourceAllocationRawResponse{}
+	if err := m.db.QueryProto(
+		"allocation_raw", &resp.ResourceEntry, afterDate.UTC(), beforeDate.UTC(),
+	); err != nil {
+		return errors.Wrap(err, "error fetching allocation data")
+	}
+
+	c.Response().Header().Set("Content-Type", "text/csv")
+	c.Response().WriteHeader(http.StatusOK)
+	labelEscaper := strings.NewReplacer("\\", "\\\\", "|", "\\|")
+
+	formatTimestamp := func(ts *timestamppb.Timestamp) string {
+		t := time.Unix(ts.Seconds, int64(ts.Nanos)).UTC()
+		return t.Format(time.RFC3339Nano)
+	}
+
+	header := "experiment_id,kind,username,labels,slots,start_time,end_time,seconds\n"
+	if _, err := c.Response().Write([]byte(header)); err != nil {
+		return err
+	}
+
+	for _, entry := range resp.ResourceEntry {
+		var labels []string
+		for _, label := range entry.Labels {
+			labels = append(labels, labelEscaper.Replace(label))
+		}
+		if _, err := fmt.Fprintf(
+			c.Response(), "%d,%s,%s,%s,%d,%s,%s,%f\n",
+			entry.ExperimentId, entry.Kind, entry.Username, strings.Join(labels, "|"), entry.Slots,
+			formatTimestamp(entry.StartTime), formatTimestamp(entry.EndTime), entry.Seconds,
+		); err != nil {
+			return err
+		}
+		c.Response().Flush()
+	}
+	return nil
```

Comment:
any reason for this vs using a `csv.NewWriter(c.Response())`. I haven't checked but I assume `c.Reponse()` is an `io.Writer`.